File: back-end/go/internal/mongodb/product_create.go
package mongodb

import (
	"context"
	"fmt"
	"github.com/icaroribeiro/hash-code-challenge/back-end/go/internal/models"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

func (d *Datastore) CreateProduct(product models.Product) (models.Product, error) {
	var result *mongo.InsertOneResult
	var err error
	var id primitive.ObjectID
	var isOK bool

	result, err = d.DB.Collection("products").InsertOne(
		context.Background(),
		product,
	)

	if err != nil {
		return product, err
	}

	// Type assertion.
	id, isOK = result.InsertedID.(primitive.ObjectID)

	if isOK {
		product.ID = id
	} else {
		return product, fmt.Errorf("it wasn't possible to obtain the id of generated document")
	}

	return product, nil
}

File: back-end/nodejs/internal/mongodb/product_create.js
var Product = require('../models/product.js');

async function CreateProduct(product, callback) {
    await Product.create(product)
        .then((doc) => {
            if (doc) {
                callback(null, {
                        id: doc._id,
                        price_in_cents: doc.price_in_cents,
                        title: doc.title,
                        description: doc.description
                    }
                )
            } else {
                callback(null, null);
            }
        })
        .catch(err => {
            callback(err, null);
        });
}

exports.CreateProduct = CreateProduct;